### DISCUSSION TITLE
Generating Ground Truth Training Set CGAL Classification Package

### DISCUSSION THREAD

**Discussion by pmff96** [CGAL/cgal#5260] [Q&A] [answered]
Not necessarily an issue, but I'd like to know whether there are any guidelines about generating ground truth training data with the specified format for the CGAL Classification package. I've been looking around for a few point cloud labelers, but none of them seem to perform what is needed for this package.

**Comment by sgiraudot** [comment, marked as answer]
There are in fact 3 layers in your question:
- what is the data structure used by the algorithms?
- how can I store it in a file?
- how can I generated

# Structure

The training set is a property map that associates, for each point, an index that corresponds to the index of the label in the `Label_set` (with `-1` meaning that the point is not labelled and thus ignored from training). This is [explained here](https://doc.cgal.org/latest/Classification/classCGAL_1_1Classification_1_1ETHZ_1_1Random__forest__classifier.html#a7c0b299cb5829f598f395bcc16df72fa).

# Storage

CGAL is generally agnostic in terms of files/IO. The easiest and most straightforward way to store this training file is to use an extensible format like PLY, and add a vertex property (for example `property int label`). [Most of the examples in the manual](https://doc.cgal.org/latest/Classification/index.html#title25) use PLY files as input. If you already have your own format, I would advise to adapt the examples to use it directly (using PLY with `property int label` is just an example and is not mandatory at all), which is how the algorithm is intended to be used.

For example, let's imagine that for some reason, you have a training set that consists in 3 separates files (ground.xyz, vegetation.xyz and buildings.xyz): in that case, you can simply load each file and fill up a training index map accordingly on the fly, without needing to merge your files and create an adapted PLY (which is also a possibility).

# Generation

Any software that lets you select and tag points can be used.

The _CGAL Polyhedron demo_ that is included with CGAL is of course the most adapted to our own packages. You can see [a video demonstration](https://www.youtube.com/watch?v=xLFm8Aw8vuY). The demo is located in [`Polyhedron/demo`](https://github.com/CGAL/cgal/tree/master/Polyhedron/demo/Polyhedron): you mainly need to compile the Classification plugin + the PLY plugin (and any other IO you may need depending on your input).

**Comment by pmff96** [comment]
Thank you for the explanation. My data does consist of a PLY file containing ground points, vegetation and power line structures, similar to the data used during the polyhedron demo. So in theory I should be able to solve my issue now, thank you once again.
